### PR TITLE
Consolidate logic from multiple projects

### PR DESCRIPTION
**Pull request by crawfxrd**
- Consolidate parsing logic that existed in here, firmware-update, and smmstore as new library functions
- Move standalone [smmstore](https://github.com/system76/smmstore) to an example in this repo
- Fix checking entries; the NULL byte was not validated
- Start adding tests

### Testing

Some unit tests have been added to verify behavior.

To test on real hardware, checkout this branch, build firmware, and replace the SMMSTORE region with a ROM that has a corrupted entry or excessive duplicates. Examples can be found in system76/firmware-open#145.

```bash
pushd apps/firmware-smmstore
git checkout consolidate
popd
./scripts/build.sh <model>
make -C coreboot/util/cbfstool
./coreboot/util/cbfstool/cbfstool ./build/<model>/firmware.rom write -r SMMSTORE -f <bad-smmstore.rom>
./scripts/flash.sh <model> --without-ec
sudo reboot
```

After rebooting, check the SMMSTORE:

```bash
./scripts/smmstore.sh
xxd -g 1 ./smmstore.rom | less
```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -67,3 +74,367 @@ pub unsafe fn smmstore_append(key: &[u8], val: &[u8]) -> Result<()> {
...
+}
+
+/// Count the number of duplicate entries in a raw region.
+/// Stops if a corrupted entry is encountered.
+pub fn count_duplicates(data: &[u8]) -> usize {
+    let mut kv = BTreeMap::<&[u8], &[u8]>::new();
+
+    let mut i = 0;
+    let mut duplicates = 0;
+    while i + 8 <= data.len() {
```
Comment (jackpot51) [inline src/lib.rs:158]: This could probably be turned into an iterator so it does not have to be rewritten three times

**Review by jackpot51 — APPROVED** [review]

**Comment by jacobgkau** [comment]
Following the testing steps in the PR description:
- Check out `consolidate` branch of `apps/firmware-smmstore`
- Build `cbfstool` and use it to add the bad `smmstore.rom` to the firmware rom.
- Flash and reboot.

After rebooting several times, I'm still getting no System76 logo on boot and am not able to access the UEFI menu. When I dump the current SMMSTORE, convert it to text using `xxd -g 1`, and compare that to the bad SMMSTORE with a diff tool, they're identical. Is this not supposed to happen any more/are there any other PRs this would depend on?

**Comment by crawfxrd** [comment]
Only the first boot after flashing with a bad SMMSTORE region should be have a black screen and not accept input, as edk2 errors happen before firmware-smmstore runs. But the SMMSTORE region will have been compacted and subsequent boots will show the splash and have working input.

Are your submodules up to date? coreboot needs to be `53ff179` or newer.

Are you using the SMMSTORE ROM from https://github.com/system76/firmware-open/issues/145#issuecomment-760431889?

**Comment by jacobgkau** [comment]
lemp10 also worked normally-- with this branch, the splash screen came back on the second reboot.

I built on galp5 without this branch and added the bad smmstore, and the splash screen came back on the second reboot after flashing. Built with this branch again and added the bad smmstore, it's now back to never showing a splash screen, which makes this seem like a regression.

When I flash galp5 with this branch, the first reboot after flashing seems longer than the first reboot without this branch, if that means anything. Subsequent reboots (with no splash screen) seem longer than usual, but not as long as the first reboot.

**Comment by crawfxrd** [comment]
@jacobgkau can you try checking out `HEAD~2` (f2afc17) and testing with that?

**Comment by jacobgkau** [comment]
@crawfxrd Here's the behavior I'm seeing right now:
- Checked out the current commit (a04cdf2). After flashing, I rebooted five times without the splash screen coming back. However, after I shut down and unplugged the AC adapter for ~10 seconds, the splash screen came back after another boot and reboot.
- Checked out `HEAD-2` (f2afc17). The splash screen came back on the third reboot after flashing, without needing to shut down/unplug to make the EC reset.
- Checked out the current commit again. The splash screen came back on the third reboot after flashing.
- Checked out `HEAD-2` again. The splash screen came back on the second reboot after flashing.

I'm not a fan of the inconsistency, but it seems like the splash screen eventually comes back. You said flashing the EC doesn't affect this, but would it make sense that an EC reset would affect it, or should that not matter either?

**Comment by crawfxrd** [comment]
Splash and UEFI app should _always_ be restored on the 2nd boot after flashing. If it doesn't then something is wrong. I especially don't like that this behavior was only seen on the galp5. I'd at least expect to see it on the lemp10 as well.

EC reset should not matter.

But without a unit set up to get EC+coreboot+edk2 logs I don't think I'll be able to tell what's wrong.

**Comment by crawfxrd** [comment]
If you build with [edk2-cbmem.patch.txt](https://github.com/system76/firmware-smmstore/files/6273164/edk2-cbmem.patch.txt) applied to edk2, then edk2 logs will get written to cbmem.

```
cd coreboot/util/cbmem
make
sudo ./cbmem -c > ~/cbmem.txt
```

**Review by jacobgkau — APPROVED** [review]
I have no longer been able to replicate this while trying throughout the past week on galp5, even after switching back to the same RAM and SSD as I was using earlier. Tested unplugged, plugged in, without suspend, and after suspending. I've tried checking out a `firmware-open` commit from when I was seeing the issue, but am still not seeing it now. Confirmed it's still working as expected on lemp10 and gaze15 as well. The splash screen always comes back on the second reboot.

I don't want to hold this up any more since I can't reproduce the issue anymore, but if I see it again, I'll get logs with the patch (if it's still working by then.)

**Pull request merged**